File: android/Reguerta/app/src/main/java/com/reguerta/user/data/products/FirestoreProductRepository.kt
package com.reguerta.user.data.products

import com.google.android.gms.tasks.Tasks
import com.google.firebase.Timestamp
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.SetOptions
import com.reguerta.user.data.firestore.ReguertaFirestoreCollection
import com.reguerta.user.data.firestore.ReguertaFirestoreEnvironment
import com.reguerta.user.data.firestore.ReguertaFirestorePath
import com.reguerta.user.data.firestore.toRepositoryException
import com.reguerta.user.domain.RepositoryErrorKind
import com.reguerta.user.domain.RepositoryException
import com.reguerta.user.domain.products.CommonPurchaseType
import com.reguerta.user.domain.products.Product
import com.reguerta.user.domain.products.ProductPricingMode
import com.reguerta.user.domain.products.ProductRepository
import com.reguerta.user.domain.products.ProductStockMode
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class FirestoreProductRepository(
    private val firestore: FirebaseFirestore,
    private val environment: ReguertaFirestoreEnvironment? = null,
) : ProductRepository {
    private val firestorePath = ReguertaFirestorePath(environment = environment)

    private val productsCollectionPath: String
        get() = firestorePath.collectionPath(ReguertaFirestoreCollection.PRODUCTS)

    override suspend fun getAllProducts(): List<Product> = withContext(Dispatchers.IO) {
        try {
            val snapshot = Tasks.await(
                firestore.collection(productsCollectionPath).get(),
            )
            decodeProductDocuments(snapshot.documents.map { document ->
                val data: Map<String, Any> = document.data ?: invalidProductDocument()
                document.id to data
            })
        } catch (error: Exception) {
            throw error.toRepositoryException(resource = "products")
        }
    }

    override suspend fun getProductsForVendor(vendorId: String): List<Product> = withContext(Dispatchers.IO) {
        try {
            val snapshot = Tasks.await(
                firestore.collection(productsCollectionPath)
                    .whereEqualTo("vendorId", vendorId)
                    .get(),
            )
            decodeProductDocuments(snapshot.documents.map { document ->
                val data: Map<String, Any> = document.data ?: invalidProductDocument()
                document.id to data
            })
        } catch (error: Exception) {
            throw error.toRepositoryException(resource = "products.vendor")
        }
    }

    override suspend fun upsertProduct(product: Product): Product = withContext(Dispatchers.IO) {
        val documentId = product.id.ifBlank {
            firestore.collection(productsCollectionPath).document().id
        }
        val persisted = product.copy(id = documentId)
        val payload = productUpsertPayload(persisted)

        try {
            Tasks.await(
                firestore.collection(productsCollectionPath)
                    .document(documentId)
                    .set(payload, SetOptions.merge()),
            )
            persisted
        } catch (error: Exception) {
            throw error.toRepositoryException(resource = "products.write")
        }
    }
}

internal fun productUpsertPayload(product: Product): Map<String, Any> = mutableMapOf<String, Any>(
    "vendorId" to product.vendorId,
    "companyName" to product.companyName,
    "name" to product.name,
    "description" to product.description,
    "productImageUrl" to (product.productImageUrl ?: FieldValue.delete()),
    "price" to product.price,
    "pricingMode" to product.pricingMode.toWireValue(),
    "unitName" to product.unitName,
    "unitAbbreviation" to (product.unitAbbreviation ?: FieldValue.delete()),
    "unitPlural" to product.unitPlural,
    "unitQty" to product.unitQty,
    "packContainerName" to (product.packContainerName ?: FieldValue.delete()),
    "packContainerAbbreviation" to (product.packContainerAbbreviation ?: FieldValue.delete()),
    "packContainerPlural" to (product.packContainerPlural ?: FieldValue.delete()),
    "packContainerQty" to (product.packContainerQty ?: FieldValue.delete()),
    "isAvailable" to product.isAvailable,
    "stockMode" to product.stockMode.toWireValue(),
    "stockQty" to (product.stockQty ?: FieldValue.delete()),
    "isEcoBasket" to product.isEcoBasket,
    "isCommonPurchase" to product.isCommonPurchase,
    "commonPurchaseType" to (product.commonPurchaseType?.toWireValue() ?: FieldValue.delete()),
    "archived" to product.archived,
    "createdAt" to product.createdAtMillis.toTimestamp(),
    "updatedAt" to product.updatedAtMillis.toTimestamp(),
    "weightStep" to (product.weightStep ?: FieldValue.delete()),
    "minWeight" to (product.minWeight ?: FieldValue.delete()),
    "maxWeight" to (product.maxWeight ?: FieldValue.delete()),
)

private fun Long.toTimestamp() = Timestamp(this / 1_000, ((this % 1_000) * 1_000_000).toInt())

internal fun decodeProductDocuments(documents: List<Pair<String, Map<String, Any>>>): List<Product> =
    documents
        .map { (documentId, data) -> decodeProductDocument(documentId, data) }
        .sortedWith(compareBy<Product> { it.archived }.thenBy { it.name.lowercase() })

internal fun decodeProductDocument(documentId: String, data: Map<String, Any?>): Product {
    val product = Product(
        id = documentId,
        vendorId = data.requiredString("vendorId"),
        companyName = data.requiredString("companyName"),
        name = data.requiredString("name"),
        description = data.optionalString("description").orEmpty(),
        productImageUrl = data.optionalString("productImageUrl"),
        price = data.requiredPositiveDouble("price"),
        pricingMode = data.optionalEnumString("pricingMode").toProductPricingMode(),
        unitName = data.requiredString("unitName"),
        unitAbbreviation = data.optionalString("unitAbbreviation"),
        unitPlural = data.requiredString("unitPlural"),
        unitQty = data.requiredPositiveDouble("unitQty"),
        packContainerName = data.optionalString("packContainerName"),
        packContainerAbbreviation = data.optionalString("packContainerAbbreviation"),
        packContainerPlural = data.optionalString("packContainerPlural"),
        packContainerQty = data.optionalPositiveDouble("packContainerQty"),
        isAvailable = data.optionalBoolean("isAvailable", default = true),
        stockMode = data.optionalEnumString("stockMode").toProductStockMode(),
        stockQty = data.optionalNonNegativeDouble("stockQty"),
        isEcoBasket = data.optionalBoolean("isEcoBasket", default = false),
        isCommonPurchase = data.optionalBoolean("isCommonPurchase", default = false),
        commonPurchaseType = data.optionalEnumString("commonPurchaseType").toCommonPurchaseType(),
        archived = data.optionalBoolean("archived", default = false),
        createdAtMillis = data.optionalTimestampMillis("createdAt"),
        updatedAtMillis = data.optionalTimestampMillis("updatedAt"),
        weightStep = data.optionalPositiveDouble("weightStep"),
        minWeight = data.optionalPositiveDouble("minWeight"),
        maxWeight = data.optionalPositiveDouble("maxWeight"),
    )
    validateSelectionRange(product)
    return product
}

private fun validateSelectionRange(product: Product) {
    if (product.pricingMode != ProductPricingMode.WEIGHT) return
    val step = product.weightStep ?: product.unitQty
    val minimumCount = kotlin.math.ceil((product.minWeight ?: step) / step)
    if (!minimumCount.isFinite() || minimumCount < 1.0 || minimumCount > Int.MAX_VALUE.toDouble()) {
        invalidProductDocument()
    }
    val maxWeight = product.maxWeight ?: return
    val maximumCount = kotlin.math.floor(maxWeight / step)
    if (!maximumCount.isFinite() ||
        maximumCount < minimumCount ||
        maximumCount > Int.MAX_VALUE.toDouble()
    ) {
        invalidProductDocument()
    }
}

private fun String?.toProductPricingMode(): ProductPricingMode = when (this?.trim()) {
    null -> ProductPricingMode.FIXED
    "fixed" -> ProductPricingMode.FIXED
    "weight" -> ProductPricingMode.WEIGHT
    else -> invalidProductDocument()
}

private fun String?.toProductStockMode(): ProductStockMode = when (this?.trim()) {
    null -> ProductStockMode.INFINITE
    "infinite" -> ProductStockMode.INFINITE
    "finite" -> ProductStockMode.FINITE
    else -> invalidProductDocument()
}

private fun String?.toCommonPurchaseType(): CommonPurchaseType? = when (this?.trim()) {
    "seasonal" -> CommonPurchaseType.SEASONAL
    "spot" -> CommonPurchaseType.SPOT
    null -> null
    else -> invalidProductDocument()
}

private fun Map<String, Any?>.requiredString(field: String): String =
    optionalString(field) ?: invalidProductDocument()

private fun Map<String, Any?>.optionalString(field: String): String? {
    val value = this[field] ?: return null
    if (value !is String) invalidProductDocument()
    return value.trim().ifBlank { null }
}

private fun Map<String, Any?>.optionalEnumString(field: String): String? {
    val value = this[field] ?: return null
    if (value !is String) invalidProductDocument()
    return value.trim().ifBlank { invalidProductDocument() }
}

private fun Map<String, Any?>.requiredDouble(field: String): Double =
    optionalDouble(field) ?: invalidProductDocument()

private fun Map<String, Any?>.requiredPositiveDouble(field: String): Double =
    requiredDouble(field).takeIf { it > 0.0 } ?: invalidProductDocument()

private fun Map<String, Any?>.optionalDouble(field: String): Double? {
    val value = this[field] ?: return null
    if (value !is Number) invalidProductDocument()
    return value.toDouble().takeIf(Double::isFinite) ?: invalidProductDocument()
}

private fun Map<String, Any?>.optionalPositiveDouble(field: String): Double? =
    optionalDouble(field)?.takeIf { it > 0.0 } ?: if (containsKey(field) && this[field] != null) {
        invalidProductDocument()
    } else {
        null
    }

private fun Map<String, Any?>.optionalNonNegativeDouble(field: String): Double? =
    optionalDouble(field)?.takeIf { it >= 0.0 } ?: if (containsKey(field) && this[field] != null) {
        invalidProductDocument()
    } else {
        null
    }

private fun Map<String, Any?>.optionalBoolean(field: String, default: Boolean): Boolean {
    val value = this[field] ?: return default
    if (value !is Boolean) invalidProductDocument()
    return value
}

private fun Map<String, Any?>.optionalTimestampMillis(field: String): Long {
    val value = this[field] ?: return 0L
    if (value !is Timestamp) invalidProductDocument()
    return value.toDate().time
}

private fun invalidProductDocument(): Nothing = throw RepositoryException(
    kind = RepositoryErrorKind.INVALID_DATA,
    resource = "products.document",
)

private fun ProductPricingMode.toWireValue(): String = when (this) {
    ProductPricingMode.FIXED -> "fixed"
    ProductPricingMode.WEIGHT -> "weight"
}

private fun ProductStockMode.toWireValue(): String = when (this) {
    ProductStockMode.FINITE -> "finite"
    ProductStockMode.INFINITE -> "infinite"
}

private fun CommonPurchaseType.toWireValue(): String = when (this) {
    CommonPurchaseType.SEASONAL -> "seasonal"
    CommonPurchaseType.SPOT -> "spot"
}

File: ios/Reguerta/Reguerta/Data/Products/FirestoreProductRepository.swift
import FirebaseFirestore
import Foundation

final class FirestoreProductRepository: @unchecked Sendable, ProductRepository {
    private let db: Firestore
    private let environment: ReguertaFirestoreEnvironment?

    init(db: Firestore = Firestore.firestore(), environment: ReguertaFirestoreEnvironment? = nil) {
        self.db = db
        self.environment = environment
    }

    private var productsCollection: CollectionReference {
        db.reguertaCollection(.products, environment: environment)
    }

    func allProducts() async throws -> [Product] {
        do {
            let snapshot = try await productsCollection.getDocuments()
            return try Self.products(from: snapshot.documents)
        } catch {
            throw FirestoreRepositoryErrorMapper.map(error, resource: "products")
        }
    }

    func products(vendorId: String) async throws -> [Product] {
        do {
            let snapshot = try await productsCollection
                .whereField("vendorId", isEqualTo: vendorId)
                .getDocuments()
            return try Self.products(from: snapshot.documents)
        } catch {
            throw FirestoreRepositoryErrorMapper.map(error, resource: "products.vendor")
        }
    }

    func upsert(product: Product) async throws -> Product {
        let documentId = product.id.isEmpty ? productsCollection.document().documentID : product.id
        let persisted = persistedProduct(from: product, with: documentId)

        do {
            try await productsCollection.document(documentId).setData(
                Self.upsertPayload(for: persisted),
                merge: true
            )
        } catch {
            throw FirestoreRepositoryErrorMapper.map(error, resource: "products.write")
        }
        return persisted
    }

    private func persistedProduct(from product: Product, with documentId: String) -> Product {
        Product(
            id: documentId,
            vendorId: product.vendorId,
            companyName: product.companyName,
            name: product.name,
            description: product.description,
            productImageUrl: product.productImageUrl,
            price: product.price,
            pricingMode: product.pricingMode,
            unitName: product.unitName,
            unitAbbreviation: product.unitAbbreviation,
            unitPlural: product.unitPlural,
            unitQty: product.unitQty,
            packContainerName: product.packContainerName,
            packContainerAbbreviation: product.packContainerAbbreviation,
            packContainerPlural: product.packContainerPlural,
            packContainerQty: product.packContainerQty,
            isAvailable: product.isAvailable,
            stockMode: product.stockMode,
            stockQty: product.stockQty,
            isEcoBasket: product.isEcoBasket,
            isCommonPurchase: product.isCommonPurchase,
            commonPurchaseType: product.commonPurchaseType,
            archived: product.archived,
            createdAtMillis: product.createdAtMillis,
            updatedAtMillis: product.updatedAtMillis
        )
    }

    static func upsertPayload(for product: Product) -> [String: Any] {
        [
            "vendorId": product.vendorId,
            "companyName": product.companyName,
            "name": product.name,
            "description": product.description,
            "productImageUrl": firestoreValue(product.productImageUrl),
            "price": product.price,
            "pricingMode": product.pricingMode.rawValue,
            "unitName": product.unitName,
            "unitAbbreviation": firestoreValue(product.unitAbbreviation),
            "unitPlural": product.unitPlural,
            "unitQty": product.unitQty,
            "packContainerName": firestoreValue(product.packContainerName),
            "packContainerAbbreviation": firestoreValue(product.packContainerAbbreviation),
            "packContainerPlural": firestoreValue(product.packContainerPlural),
            "packContainerQty": firestoreValue(product.packContainerQty),
            "isAvailable": product.isAvailable,
            "stockMode": product.stockMode.rawValue,
            "stockQty": firestoreValue(product.stockQty),
            "isEcoBasket": product.isEcoBasket,
            "isCommonPurchase": product.isCommonPurchase,
            "commonPurchaseType": firestoreValue(product.commonPurchaseType?.rawValue),
            "archived": product.archived,
            "createdAt": timestamp(for: product.createdAtMillis),
            "updatedAt": timestamp(for: product.updatedAtMillis),
            "weightStep": firestoreValue(product.weightStep),
            "minWeight": firestoreValue(product.minWeight),
            "maxWeight": firestoreValue(product.maxWeight)
        ]
    }

    private static func timestamp(for millis: Int64) -> Timestamp {
        Timestamp(date: Date(timeIntervalSince1970: TimeInterval(millis) / 1_000))
    }

    private static func firestoreValue<Value>(_ value: Value?) -> Any {
        if let value {
            return value
        }
        return FieldValue.delete()
    }

    private static func sortProducts(_ lhs: Product, _ rhs: Product) -> Bool {
        if lhs.archived != rhs.archived {
            return !lhs.archived && rhs.archived
        }
        return lhs.name.localizedCaseInsensitiveCompare(rhs.name) == .orderedAscending
    }

    private static func products(from documents: [QueryDocumentSnapshot]) throws -> [Product] {
        try documents
            .map { document in
                try product(documentID: document.documentID, data: document.data())
            }
            .sorted(by: sortProducts)
    }

    static func product(documentID: String, data: [String: Any]) throws -> Product {
        let product = Product(
            id: documentID,
            vendorId: try requiredString(data, field: "vendorId"),
            companyName: try requiredString(data, field: "companyName"),
            name: try requiredString(data, field: "name"),
            description: try optionalString(data, field: "description") ?? "",
            productImageUrl: try optionalString(data, field: "productImageUrl"),
            price: try requiredPositiveDouble(data, field: "price"),
            pricingMode: try pricingMode(data),
            unitName: try requiredString(data, field: "unitName"),
            unitAbbreviation: try optionalString(data, field: "unitAbbreviation"),
            unitPlural: try requiredString(data, field: "unitPlural"),
            unitQty: try requiredPositiveDouble(data, field: "unitQty"),
            packContainerName: try optionalString(data, field: "packContainerName"),
            packContainerAbbreviation: try optionalString(data, field: "packContainerAbbreviation"),
            packContainerPlural: try optionalString(data, field: "packContainerPlural"),
            packContainerQty: try optionalPositiveDouble(data, field: "packContainerQty"),
            isAvailable: try optionalBool(data, field: "isAvailable", default: true),
            stockMode: try stockMode(data),
            stockQty: try optionalNonNegativeDouble(data, field: "stockQty"),
            isEcoBasket: try optionalBool(data, field: "isEcoBasket", default: false),
            isCommonPurchase: try optionalBool(data, field: "isCommonPurchase", default: false),
            commonPurchaseType: try commonPurchaseType(data),
            archived: try optionalBool(data, field: "archived", default: false),
            createdAtMillis: try optionalTimestampMillis(data, field: "createdAt"),
            updatedAtMillis: try optionalTimestampMillis(data, field: "updatedAt"),
            weightStep: try optionalPositiveDouble(data, field: "weightStep"),
            minWeight: try optionalPositiveDouble(data, field: "minWeight"),
            maxWeight: try optionalPositiveDouble(data, field: "maxWeight")
        )
        try validateSelectionRange(product)
        return product
    }

    private static func requiredString(_ data: [String: Any], field: String) throws -> String {
        guard let value = try optionalString(data, field: field) else { throw invalidDocumentError }
        return value
    }

    private static func optionalString(_ data: [String: Any], field: String) throws -> String? {
        guard let value = data[field] else { return nil }
        if value is NSNull { return nil }
        guard let string = value as? String else { throw invalidDocumentError }
        let trimmed = string.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }

    private static func requiredDouble(_ data: [String: Any], field: String) throws -> Double {
        guard let value = try optionalDouble(data, field: field) else { throw invalidDocumentError }
        return value
    }

    private static func requiredPositiveDouble(_ data: [String: Any], field: String) throws -> Double {
        let value = try requiredDouble(data, field: field)
        guard value > 0 else { throw invalidDocumentError }
        return value
    }

    private static func optionalDouble(_ data: [String: Any], field: String) throws -> Double? {
        guard let value = data[field] else { return nil }
        if value is NSNull { return nil }
        guard !(value is Bool), let number = value as? NSNumber else { throw invalidDocumentError }
        let double = number.doubleValue
        guard double.isFinite else { throw invalidDocumentError }
        return double
    }

    private static func optionalPositiveDouble(_ data: [String: Any], field: String) throws -> Double? {
        guard let value = try optionalDouble(data, field: field) else { return nil }
        guard value > 0 else { throw invalidDocumentError }
        return value
    }

    private static func optionalNonNegativeDouble(_ data: [String: Any], field: String) throws -> Double? {
        guard let value = try optionalDouble(data, field: field) else { return nil }
        guard value >= 0 else { throw invalidDocumentError }
        return value
    }

    private static func optionalBool(_ data: [String: Any], field: String, default defaultValue: Bool) throws -> Bool {
        guard let value = data[field] else { return defaultValue }
        if value is NSNull { return defaultValue }
        guard let bool = value as? Bool else { throw invalidDocumentError }
        return bool
    }

    private static func optionalTimestampMillis(_ data: [String: Any], field: String) throws -> Int64 {
        guard let value = data[field] else { return 0 }
        if value is NSNull { return 0 }
        guard let timestamp = value as? Timestamp else { throw invalidDocumentError }
        return Int64(timestamp.dateValue().timeIntervalSince1970 * 1_000)
    }

    private static func pricingMode(_ data: [String: Any]) throws -> ProductPricingMode {
        guard let rawValue = try optionalEnumString(data, field: "pricingMode") else { return .fixed }
        guard let value = ProductPricingMode(rawValue: rawValue) else { throw invalidDocumentError }
        return value
    }

    private static func stockMode(_ data: [String: Any]) throws -> ProductStockMode {
        guard let rawValue = try optionalEnumString(data, field: "stockMode") else { return .infinite }
        guard let value = ProductStockMode(rawValue: rawValue) else { throw invalidDocumentError }
        return value
    }

    private static func commonPurchaseType(_ data: [String: Any]) throws -> CommonPurchaseType? {
        guard let rawValue = try optionalString(data, field: "commonPurchaseType") else { return nil }
        guard let value = CommonPurchaseType(rawValue: rawValue) else { throw invalidDocumentError }
        return value
    }

    private static func optionalEnumString(_ data: [String: Any], field: String) throws -> String? {
        guard let value = data[field] else { return nil }
        if value is NSNull { return nil }
        guard let string = value as? String else { throw invalidDocumentError }
        let trimmed = string.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { throw invalidDocumentError }
        return trimmed
    }

    private static func validateSelectionRange(_ product: Product) throws {
        guard product.pricingMode == .weight else { return }
        let step = product.effectiveWeightStep
        let minimumCount = ceil((product.minWeight ?? step) / step)
        guard minimumCount.isFinite,
              minimumCount >= 1,
              minimumCount <= Double(Int32.max) else {
            throw invalidDocumentError
        }
        guard let maxWeight = product.maxWeight else { return }
        let maximumCount = floor(maxWeight / step)
        guard maximumCount.isFinite,
              maximumCount >= minimumCount,
              maximumCount <= Double(Int32.max) else {
            throw invalidDocumentError
        }
    }

    private static var invalidDocumentError: RepositoryError {
        .invalidData(resource: "products.document")
    }
}
